modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md

```diff
@@ -0,0 +1,167 @@
+---
+layout: docs
+title: Generic
+permalink: /docs/generic/coproduct/
+---
+
+## Arrow Generic
+
+{:.beginner}
+beginner
+
+`arrow-generic` provides meta programming facilities over Product types like data classes, tuples, and heterogeneous lists; and Coproduct types like sealed classes.
+
+### Install
+
+```groovy
+compile 'io.arrow-kt:arrow-generic:$arrow_version'
+```
+
+### Features
+
+#### Coproduct
+
+Coproducts represent a container in which only one of the specified set of types exist. It's very similar in `Either` in that respect. `Either` supports one of two values, an `Either<A, B>` has to contain an instance of `A` or `B`. We can extrapolate that concept to `N` number of types. So a `Coproduct5<A, B, C, D, E>` has to contain an instance of `A`, `B`, `C`, `D`, or `E`. For example, perhaps there's a search function for a Car Dealer app that can show `Dealership`s, `Car`s and `SalesPerson`s, we could model that list of results as `List<Coproduct3<Dealership, Car, SalesPerson>`. The result would contain a list of heterogeneous elements but each element is one of `Dealership`, `Car` or `SalesPerson` and our UI can render the list elements based on those types.
+
+Let's say we have an api. Our api operates under the following conditions:
+- Every endpoint could tell the client, `ServerError`, `UserUnauthorized` or `OverRequestLimit`.
+- For specific endpoints, we may have specific results to the context of what they do, for example, an endpoint for registering a `Car` for some service. This endpoint could return to us `CarAlreadyRegistered`, `StolenCar` or `SuccessfullyRegistered`. For fun, let's say the `SuccessfullyRegistered` response also contains a `Registration` object with some data.
+
+Because we have some common errors that every endpoint can return to us, we can define a sealed class of these and call them `CommonServerError` because they make sense to be sealed together for reusability. Likewise, we can logically group our specific errors into `RegistrationError` and we have `Registration` as our success type.
+
+With Coproducts, we're able to define a result for this api call as `Coproduct3<CommonServerError, RegistrationError, Registration>`. We've been able to compose these results without having to write our own sealed class containing all the common errors for each endpoint.
+
+#### Extensions
+
+##### coproductOf
+
+So now that we've got our api response modeled, we need to be able to create an instance of `Coproduct3`.
+
+```kotlin:ank
+import arrow.generic.*
+import arrow.generic.coproduct3.coproductOf
+
+val apiResult = coproductOf<CommonServerError, RegistrationError, Registration>(ServerError) //Returns Coproduct3<CommonServerError, RegistrationError, Registration>
+```
+
+There are `coproductOf` constructor functions for each Coproduct. All we have to do is pass in our value and have the correct type parameters on it, the value must be a type declared on the function call.
```

Comment:
First line, I'd better say "...for each Coproduct regardless of the arity (number of types)."